docs/architecture/types/README.md

```diff
@@ -0,0 +1,345 @@
+corresponding to the instantiation. A resource's
+[*`Type`*](https://www.pulumi.com/docs/iac/concepts/resources/names/#types)
+(often captured in a [schema](schema) in the case of a custom provider resource)
+specifies a set of [*input properties*](inputs) that define the resource's
+desired state, and a set of [*output properties*](outputs) that represent the
+last actual state that Pulumi recorded.
+
+[^urn-uniqueness]:
+    In an ideal world, URNs would be globally unique, but in practice there are
+    a few exceptions. Most notable of these is that, within a stack
+    (specifically, within a stack's [*state*](state-snapshots)), a URN may
+    appear multiple times in order to identify, for instance, both a copy of a
+    resource that is pending deletion and a copy that will be created to replace
+    that deleted instance.
+
+(outputs)=
+### Outputs
+
+`Output<T>` is perhaps the most important member of Pulumi's type system,
+representing a node in a Pulumi program graph that produces a value of type `T`.
+An `Output<T>` value behaves like a `Promise<T>` in that it represents an
+asynchronous computation that will eventually produce a value of type `T`, but
+it also carries additional metadata that describes the resources on which the
+value depends, whether the value is known or unknown, and whether or not the
+value is secret.
+
+#### Dependency tracking
+
+If an `Output<T>` value is the result of a resource operation---that is, if it
+is an output property of some resource---it is said to *depend* on that
+resource. Output dependencies are *transitive*: if an `Output<T>` value `O₁` is
+depended upon by another `Output<T>` value `O₂`, then any value `O₃` that
+depends on `O₂` also depends on `O₁`. Dependency tracking is used to ensure that
+the correct order of operations is maintained when creating or updating
+resources.
+
+#### Unknowns
+
+An `Output<T>` may be *unknown* if it depends on the result of a resource
+operation that will not be run because it is part of a `pulumi preview`.
+Previews typically produce unknowns for properties with values that cannot be
+determined until the resource is actually created or updated. If a value of type
+`Output<T>` is unknown, any computation that depends on its concrete value must
+not run, and must therefore also produce an unknown `Output<T>`.
+
+#### Secrets
+
+An `Output<T>` may be marked as *secret* if its concrete value contains
+sensitive information. If a value of type `Output<T>` is secret, the result of
+any computation that depends on its concrete value must also be secret.
+
+(inputs)=
+### Inputs
+
+The partner of `Output<T>` is `Input<T>`, which represents a value that is
+either of type `T` or `Output<T>` (and is thus defined as `Union<T,
+Output<T>>`). In this manner, input properties may accept either plain values
+defined in the program outright, or values which arise from the outputs of other
+resources.
+
+### Transformations
+
+(inputshape)=
+#### `inputShape`
```

Comment:
This is a massive improvement on the previous doc on this, I find this massively easier to read and understand.